compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Generate.hs

```diff
@@ -0,0 +1,192 @@
+combineGO :: GenOutput -> GenOutput
+          -- ^ The two generator outputs to combine.
+          -> GenOutput
+combineGO genOut1 genOut2
+  = extendGOUpds (genOut2 ^. goUpd)
+  $ extendGODelta (genOut2 ^. goDel)
+    genOut1
+
+updateGOExpr :: Expr
+             -- ^ The new output expression.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+updateGOExpr expr = set goExp expr
+
+extendGOUpds :: UpdateSet
+             -- ^ The extension of the update set.
+             -> GenOutput
+             -- ^ The current generator output.
+             -> GenOutput
+extendGOUpds upds = over goUpd (concatUpdateSet upds)
+
+extendGODelta :: Delta
+              -- ^ The additional context extension.
+              -> GenOutput
+              -- ^ The current generator output.
+              -> GenOutput
+extendGODelta delta = over goDel (concatDelta delta)
+
+genPackages :: MonadDelta m => [(PackageId, (Package, Maybe PackageName))]
+            -> m Delta
+genPackages inp = ask >>= \del0 -> foldM genPackage' del0 inp
+  where
+    -- TODO: These prime definitions are bit silly. Merge with the regular
+    -- functions.
+    -- TODO: This >>= return . concatDelta returns a lot. Figure out a way to
+    -- abstract over it. Same for this returning foldM.
+    genPackage' :: MonadDelta m => Delta
+                -> (PackageId, (Package, Maybe PackageName)) -> m Delta
+    genPackage' deli (idi, (paci, _)) =
+      -- TODO: This is getting quite unreadable.
+      foldM (\deli' modi -> genModule' deli' (PRImport idi, modi))
+            deli (NM.toList $ packageModules paci)
+      >>= return . (concatDelta deli)
+    genModule' :: MonadDelta m => Delta -> (PackageRef, Module) -> m Delta
+    genModule' deli (paci, modi) = introDelta deli (genModule paci modi)
+                                   >>= return . (concatDelta deli)
+
+-- TODO: Type synonyms and data types are ignored for now.
+genModule :: MonadDelta m => PackageRef -> Module -> m Delta
+genModule pac mod = do
+  del0 <- ask
+  del1 <- foldM genValue' del0 (NM.toList $ moduleValues mod)
+  del2 <- foldM genTemplate' del1 (NM.toList $ moduleTemplates mod)
+  return $ concatDelta del1 del2
+  where
+    genValue' :: MonadDelta m => Delta -> DefValue -> m Delta
+    genValue' deli vali = introDelta deli (genValue pac (moduleName mod) vali)
+                          >>= return . (concatDelta deli)
+    genTemplate' :: MonadDelta m => Delta -> Template -> m Delta
+    genTemplate' deli temi = introDelta deli (genTemplate temi)
+                          >>= return . (concatDelta deli)
+
+genValue :: MonadDelta m => PackageRef -> ModuleName -> DefValue -> m Delta
+genValue pac mod val = do
+  expOut <- genExpr (dvalBody val)
+  let qname = Qualified pac mod (fst $ dvalBinder val)
+  return $ set devals (singleton qname (expOut ^. goExp, expOut ^. goUpd)) emptyDelta
+
+-- TODO: Handle annotated choices, by returning a set of annotations.
```

Comment:
What is an annotated choice? Presumably that’s some new annotation that you want to add to express constraints that should be verified?